Telemetry frontend: metrics to stdout, tracing in http handlers, gRPC interceptor

gzhao408 opened the pull request:
Partially resolves #181, subtask of #132 

Progress so far:
- [x] Update docker image to be compatible with OpenTelemetry libraries
- [x] Add basic metrics to HTTP handlers
- [x] Add custom metrics (HTTP request count, HTTP errors count, HTTP request latency)
(https://github.com/GoogleCloudPlatform/stackdriver-sandbox/pull/183#issuecomment-664690690).
- [x] Add OTel gRPC interceptors to gRPC connections and the HTTP server
- [x] Export HTTP server traces to Cloud Trace

TODO:
- [ ] Testing script for tracing.
- [ ] Export metrics to Cloud Monitoring (blocked by dependency update in exporter, opened a [PR](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/73))


**How to test Metrics Manually:** Deploy (e.g. run `make deploy-continuous PROJECT_ID=<projectid>`) and observe what's printed to stdout from frontend service server. It'll look something like this, and obviously respond to requests to hipster shop.

![image](https://user-images.githubusercontent.com/25183001/88724877-11e9ad00-d0e0-11ea-87d8-32bc9870dafd.png)


<details><summary>Sample Output in Pretty JSON</summary>
<p>

```js
{

   "time":"2020-07-28T21:29:30.458881296Z",
   "updates":[

      {

         "name":"http_request_count{instrumentation.name=hipstershop/frontend}",
         "sum":5
      },
      {

         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.5.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {

         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {

         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "min":49211,
         "max":49211,
         "sum":49211,
         "count":1,
         "quantiles":[

            {

               "q":0.5,
               "v":49211
            },
            {

               "q":0.9,
               "v":49211
            },
            {

               "q":0.99,
               "v":49211
            }
         ]
      },
      {

         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {

         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "min":38423,
         "max":118995,
         "sum":157418,
         "count":2,
         "quantiles":[

            {

               "q":0.5,
               "v":118995
            },
            {

               "q":0.9,
               "v":118995
            },
            {

               "q":0.99,
               "v":118995
            }
         ]
      },
      {

         "name":"http_response_errors{instrumentation.name=hipstershop/frontend}",
         "sum":1
      },
      {

         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.5.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":4
      },
      {

         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.5.7:8080,http.scheme=http,http.server_name=frontend}",
         "min":178,
         "max":203,
         "sum":381,
         "count":2,
         "quantiles":[

            {

               "q":0.5,
               "v":203
            },
            {

               "q":0.9,
               "v":203
            },
            {

               "q":0.99,
               "v":203
            }
         ]
      },
      {

         "name":"http_request_latency{instrumentation.name=hipstershop/frontend}",
         "min":0,
         "max":119,
         "sum":207,
         "count":5,
         "quantiles":[

            {

               "q":0.5,
               "v":39
            },
            {

               "q":0.9,
               "v":119
            },
            {

               "q":0.99,
               "v":119
            }
         ]
      },
      {

         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":22279
      },
      {

         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":32
      }
   ]
}
```
</p>
</details>



**How to test Tracing Manually**:  Deploy (e.g. run `make deploy-continuous PROJECT_ID=<projectid>`) and then go to GCP Console > Cloud Trace > Trace List and filter using `LABEL:g.co/agent:opentelemetry`. You should see a number of traces with URI `span.frontend-server` with various latencies. 
![image](https://user-images.githubusercontent.com/25183001/88726686-09df3c80-d0e3-11ea-89ba-b059e6b3bc0c.png)

Choose one with a higher latency (i.e. higher on the y-axis) and you'll see something interesting like this:
![image](https://user-images.githubusercontent.com/25183001/88724410-4741cb00-d0df-11ea-9fa6-f60c7e4c5870.png)

simonz130 commented on src/frontend/handlers.go:
```diff
@@ -88,6 +92,7 @@ func (fe *frontendServer) homeHandler(w http.ResponseWriter, r *http.Request) {
 
 func (fe *frontendServer) productHandler(w http.ResponseWriter, r *http.Request) {
 	log := r.Context().Value(ctxKeyLog{}).(logrus.FieldLogger)
+	trace.SpanFromContext(r.Context()).SetAttributes(kv.String("name", "product"))
```
Is this required?
Can we "hide" it by using interceptors for all http incoming requests?

gzhao408 replied:
hidden

simonz130 commented on src/frontend/main.go:
```diff
@@ -31,6 +31,18 @@ import (
 	"go.opencensus.io/stats/view"
 	"go.opencensus.io/trace"
+
+	texporter "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace"
```
Worth adding a comment to explain which package are you using (exporter directly vs. collector and then exporting it)

gzhao408 replied:
done

simonz130 commented on src/frontend/main.go:
```diff
@@ -157,6 +184,17 @@ func initStats(log logrus.FieldLogger, exporter *stackdriver.Exporter) {
 	}
 }
 
+func initMeter(log logrus.FieldLogger) *push.Controller {
```
document the method
also perhaps worth renaming to initMetricsCollection or something along those lines so it's clear what it does.

gzhao408 replied:
done

simonz130 commented on src/frontend/main.go:
```diff
@@ -157,6 +184,17 @@ func initStats(log logrus.FieldLogger, exporter *stackdriver.Exporter) {
+	return pusher
+}
+
 func initStackdriverTracing(log logrus.FieldLogger) {
```
This method name is not correct anymore :)

gzhao408 replied:
changed to `initOpenCensus` :)

simonz130 commented on src/frontend/main.go:
```diff
@@ -187,6 +225,23 @@ func initTracing(log logrus.FieldLogger) {
+	tp, err := sdktrace.NewProvider(sdktrace.WithConfig(
+		sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.New(standard.ServiceNameKey.String("FrontendService"))))
```
Isn't the package picking up the resource automatically? It should but it's possible that it's not implemented yet.
If it's not implemented yet, worth changing this to be "GKE". The resource name is supposed to be a platform on which the code is running.

gzhao408 replied:
Not 100% sure, but I changed it to GKE

simonz130 commented on src/frontend/main.go:
```diff
@@ -225,7 +280,9 @@ func mustConnGRPC(ctx context.Context, conn **grpc.ClientConn, addr string) {
-		grpc.WithStatsHandler(&ocgrpc.ClientHandler{}))
+		grpc.WithStatsHandler(&ocgrpc.ClientHandler{}),
+		grpc.WithUnaryInterceptor(grpctrace.UnaryClientInterceptor(global.TraceProvider().Tracer("frontend"))),
+		grpc.WithStreamInterceptor(grpctrace.StreamClientInterceptor(global.TraceProvider().Tracer("frontend"))))
```
I see you use interceptor. There should be a way to detect the name of the grpc action that is being called, so you shouldn't need to call 
```
trace.SpanFromContext(r.Context()).SetAttributes(kv.String("name", "home"))
```
and similar in each method

gzhao408 replied:
removed and added it to `ServeHTTP` with `r.URL.Path` instead

simonz130 commented on src/frontend/middleware.go:
```diff
@@ -58,6 +58,8 @@ func (lh *logHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	requestID, _ := uuid.NewRandom()
 	ctx = context.WithValue(ctx, ctxKeyRequestID{}, requestID.String())
 
+	http_request_count.Add(ctx, 1)
```
There should be a built-in metric for instrumenting grpc with opentelemetry that should capture this. If this doesn't exist, then it's fine to leave it like this for now. Perhaps worth adding a //todo comment to remove it in the future when such metric is implemented in the core OT library.

gzhao408 replied:
There isn't a Views API and this isn't collected automatically, so it's custom metrics for now. With `othttp` it has some automatic ones but not the exact ones we want.

simonz130 commented on src/frontend/main.go:
```diff
@@ -186,7 +224,24 @@ func initTracing(log logrus.FieldLogger) {
 	// trace.ProbabilitySampler set at the desired probability.
 	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
-	initStackdriverTracing(log)
+	initOpenCensus(log)
```
are you making a recursive call here?

gzhao408 replied:
no - `main` calls `initTracing` which calls `initOpenCensus`

gzhao408 commented:
This is what I can see in the Cloud Trace console:

![image](https://user-images.githubusercontent.com/25183001/88601931-8fe97d80-d026-11ea-9e25-3bba00aee692.png)

I think this is the most I can get right now since I haven't instrumented any of the other services with OTel

simonz130 commented on src/frontend/main.go:
```diff
@@ -186,7 +224,24 @@ func initTracing(log logrus.FieldLogger) {
-	initStackdriverTracing(log)
+	initOpenCensus(log)
+
+	// Initialize OpenTelemetry Tracer
```
I don't think the comment is right. You are initializing the exporter to Cloud Tracing (Google product, former Stackdriver Tracing)

gzhao408 replied:
sorry yeah, updated

gzhao408 commented:
Example of metrics stdout (includes othttp and the custom metrics):
<details><summary>Output</summary>
<p>

```js
{
   "time":"2020-07-27T23:18:24.31340596Z",
   "updates":[
      {
         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.3.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {
         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.3.7:8080,http.scheme=http,http.server_name=frontend}",
         "min":185,
         "max":273,
         "sum":458,
         "count":2,
         "quantiles":[
            {
               "q":0.5,
               "v":273
            },
            {
               "q":0.9,
               "v":273
            },
            {
               "q":0.99,
               "v":273
            }
         ]
      },
      {
         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":30009
      },
      {
         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {
         "name":"http.server.response_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=10.60.3.7:8080,http.scheme=http,http.server_name=frontend}",
         "sum":4
      },
      {
         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "min":7628,
         "max":26039,
         "sum":54705,
         "count":3,
         "quantiles":[
            {
               "q":0.5,
               "v":21038
            },
            {
               "q":0.9,
               "v":26039
            },
            {
               "q":0.99,
               "v":26039
            }
         ]
      },
      {
         "name":"http_request_latency{instrumentation.name=hipstershop/frontend}",
         "min":0,
         "max":26,
         "sum":59,
         "count":6,
         "quantiles":[
            {
               "q":0.5,
               "v":7
            },
            {
               "q":0.9,
               "v":26
            },
            {
               "q":0.99,
               "v":26
            }
         ]
      },
      {
         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "sum":32
      },
      {
         "name":"http.server.duration{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.request_content_length=32,http.scheme=http,http.server_name=frontend}",
         "min":4430,
         "max":4430,
         "sum":4430,
         "count":1,
         "quantiles":[
            {
               "q":0.5,
               "v":4430
            },
            {
               "q":0.9,
               "v":4430
            },
            {
               "q":0.99,
               "v":4430
            }
         ]
      },
      {
         "name":"http_request_count{instrumentation.name=hipstershop/frontend}",
         "sum":6
      },
      {
         "name":"http.server.request_content_length{instrumentation.name=go.opentelemetry.io/otel/instrumentation/othttp,http.flavor=1.1,http.host=34.83.167.77,http.scheme=http,http.server_name=frontend}",
         "sum":0
      },
      {
         "name":"http_response_errors{instrumentation.name=hipstershop/frontend}",
         "sum":2
      }
   ]
}
```
</p>
</details>

simonz130 commented on src/frontend/main.go:
```diff
@@ -186,7 +224,24 @@ func initTracing(log logrus.FieldLogger) {
+	tp, err := sdktrace.NewProvider(sdktrace.WithConfig(
+		sdktrace.Config{DefaultSampler: sdktrace.AlwaysSample()}),
+		sdktrace.WithSyncer(exporter),
+		sdktrace.WithResource(resource.New(standard.ServiceNameKey.String("GKE"))))
```
Do you know if they support resource auto-detection? Or at least a way to pass a pre-defined constant for GKE?
If both are not an option, at least please add a comment with TODO so when resource autodetection is supported, we should fix it here.

gzhao408 replied:
added a TODO comment
there's no no predefined constant in the `standard` api
I'm 90% sure no auto-detection

gzhao408 replied:
u were right I didn't need it :P

simonz130 commented on src/frontend/middleware.go:
```diff
@@ -74,6 +81,12 @@ func (lh *logHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			"http.resp.took_ms": int64(time.Since(start) / time.Millisecond),
 			"http.resp.status":  rr.status,
 			"http.resp.bytes":   rr.b}).Debugf("request complete")
+		// TODO: remove when default metrics through Views API are avilable in OpenTelemetry
```
nit: typo - available

gzhao408 commented:
Rebased and squashed into 3 commits: docker image update, metrics + export to stdout, tracing + export to cloud trace
Also added some instruction on how to test manually in the PR description
Ready for review again!
